### PR TITLE
Npp754

### PR DESCRIPTION
**Pull request by ClaudiaFrank**
2nd attempt - added missing exception classes

### CODE REVIEW

File: PythonScript/src/NotepadPlusWrapper.cpp [hunk 1]
```diff
@@ -900,6 +917,162 @@ boost::python::object NotepadPlusWrapper::allocateMarker(int quantity)
...
+}
+
+bool NotepadPlusWrapper::isStatusBarHidden()
+{
+	return callNotepad(NPPM_ISSTATUSBARHIDDEN);
+}
+
+void NotepadPlusWrapper::saveFile(boost::python::str filename)
+{
+	std::shared_ptr<TCHAR> s = WcharMbcsConverter::char2tchar((const char*)boost::python::extract<const char*>(filename));
```
Comment (chcg) [inline PythonScript/src/NotepadPlusWrapper.cpp]: Is this behaving correctly with unicode file names?

Comment (ClaudiaFrank) [reply PythonScript/src/NotepadPlusWrapper.cpp]: Yes, I can open, activate, reload, save etc... without a problem but
notepad.launchFindInFilesDlg seems to be behaving different.

File: PythonScript/src/NotepadPlusWrapper.h [hunk 2]
```diff
@@ -565,36 +739,28 @@ class NotepadPlusWrapper
...
 	HINSTANCE m_hInst;
-	
+
 	callbackT m_callbacks;
 	bool m_notificationsEnabled;
 	HANDLE m_callbackMutex;
 
-    void notAllowedInScintillaCallback(const char *message);
+	void notAllowedInScintillaCallback(const char *message);
+	void NotepadPlusWrapper::invalidValueProvided(const char *message);
```
Comment (chcg) [inline PythonScript/src/NotepadPlusWrapper.h]: NotepadPlusWrapper:: is not necessary here.

File: PythonScript/src/NotepadPlusWrapper.cpp [hunk 3]
```diff
@@ -591,7 +603,14 @@ BufferEncoding NotepadPlusWrapper::getEncoding()
 
 BufferEncoding NotepadPlusWrapper::getBufferEncoding(intptr_t bufferID)
 {
-	return static_cast<BufferEncoding>(callNotepad(NPPM_GETBUFFERENCODING, static_cast<WPARAM>(bufferID)));
+	if (bufferID >= 0)
```
Comment (chcg) [inline PythonScript/src/NotepadPlusWrapper.cpp]: Do you have an example which shows the problem with bufferID < 0? Seems negative values are ok. The bufferID is directly used as pointer. There is just no check in N++ that the bufferID is a valid one and therefore an invalid pointer access could happen if the bufferID is unknown.

Comment (ClaudiaFrank) [reply PythonScript/src/NotepadPlusWrapper.cpp]: For example notepad.reloadBuffer(-1,True) or notepad.getLangType(-1)

Comment (chcg) [reply PythonScript/src/NotepadPlusWrapper.cpp]: Callstack is:
 
 notepad++.exe!Buffer::setDeferredReload() Zeile 465	C++
>	notepad++.exe!FileManager::reloadBufferDeferred(Buffer * id) Zeile 681	C++
 	notepad++.exe!Notepad_plus::doReload(Buffer * id, bool alert) Zeile 448	C++
 	notepad++.exe!Notepad_plus::process(HWND__ * hwnd, unsigned int message, unsigned __int64 wParam, __int64 lParam) Zeile 411	C++
 	notepad++.exe!Notepad_plus_Window::runProc(HWND__ * hwnd, unsigned int message, unsigned __int64 wParam, __int64 lParam) Zeile 120	C++
 	notepad++.exe!Notepad_plus_Window::Notepad_plus_Proc(HWND__ * hwnd, unsigned int message, unsigned __int64 wParam, __int64 lParam) Zeile 96	C++

and FileManager::reloadBufferDeferred(BufferID id) at
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/3a52ab1af16df8833eaab293106a85ba909be864/PowerEditor/src/ScitillaComponent/Buffer.cpp#L677

is not checking if the provided bufferID is a valid one. So I guess that needs to be done somewhere at the wrapper level. 
Similar to https://github.com/notepad-plus-plus/notepad-plus-plus/blob/3a52ab1af16df8833eaab293106a85ba909be864/PowerEditor/src/ScitillaComponent/Buffer.cpp#L506

**Comment by chcg** [comment]
@ClaudiaFrank I merged your changes and will work on the remaining issues observed.

File: PythonScript/src/NotepadPython.cpp [hunk 4]
```diff
@@ -35,7 +38,7 @@ void export_notepad()
 		.def("destroyScintilla", &NotepadPlusWrapper::destroyScintilla, boost::python::args("editor"), "Destroy a Scintilla handle created with createScintilla.")
 		.def("getCurrentDocIndex", &NotepadPlusWrapper::getCurrentDocIndex, boost::python::args("view"), "Gets the current active index for the given view (0 or 1)")
 		.def("setStatusBar", &NotepadPlusWrapper::setStatusBar, boost::python::args("section", "text"), "Sets the status bar text. Call with the STATUSBARSECTION, and the string to show")
-		.def("getPluginMenuHandle", &NotepadPlusWrapper::getPluginMenuHandle, "Gets the handle for the Plugins menu.")
+		.def("getPluginMenuHandle", &NotepadPlusWrapper::getPluginMenuHandle, "DEPRECATED Gets the handle for the Plugins menu.")
```
Comment (chcg) [inline PythonScript/src/NotepadPython.cpp]: @ClaudiaFrank What is the replacement? getMenuHandle?

File: PythonScript/src/NotepadPlusWrapper.cpp [hunk 5]
```diff
@@ -89,18 +89,27 @@ void NotepadPlusWrapper::notify(SCNotification *notifyCode)
...
 		case NPPN_SHUTDOWN:
 		case NPPN_TBMODIFICATION:  // I hope no-one ever has to use this!
+		case NPPN_BEFORESHUTDOWN:
+		case NPPN_CANCELSHUTDOWN:
 			// No additional parameters
 			break;
 
 		case NPPN_READONLYCHANGED:
-			params["bufferID"] = notifyCode->nmhdr.hwndFrom;
+			params["bufferID"] = (unsigned int)notifyCode->nmhdr.hwndFrom;
```
Comment (chcg) [inline PythonScript/src/NotepadPlusWrapper.cpp]: Why is this cast be needed? Currently I think it is not necessary.

File: PythonScript/src/NotepadPlusWrapper.cpp [hunk 6]
```diff
@@ -118,11 +127,15 @@ void NotepadPlusWrapper::notify(SCNotification *notifyCode)
 			}
 			break;
 
-		
+		case NPPN_DOCORDERCHANGED:
+			params["newIndex"] = (unsigned int)notifyCode->nmhdr.hwndFrom;
```
Comment (chcg) [inline PythonScript/src/NotepadPlusWrapper.cpp]: Why is this cast be needed? Currently I think it is not necessary.

File: PythonScript/src/NotepadPlusWrapper.cpp [hunk 7]
```diff
@@ -118,11 +127,15 @@ void NotepadPlusWrapper::notify(SCNotification *notifyCode)
...
+		case NPPN_DOCORDERCHANGED:
+			params["newIndex"] = (unsigned int)notifyCode->nmhdr.hwndFrom;
+			params["bufferID"] = notifyCode->nmhdr.idFrom;
+			break;
+
 		default:
 			// Unknown notification, so just fill in the parameters as integers.
 			params["idFrom"] = notifyCode->nmhdr.idFrom;
-			params["hwndFrom"] = notifyCode->nmhdr.hwndFrom;
+			params["hwndFrom"] = (unsigned int)notifyCode->nmhdr.hwndFrom;
```
Comment (chcg) [inline PythonScript/src/NotepadPlusWrapper.cpp]: Why is this cast be needed? Currently I think it is not necessary.

**Pull request merged**